eth_utils/abi.py

```diff
@@ -1,68 +1,802 @@
+    """
+    Return a list of each ``ABIElement`` that is of type ``type``.
+
+    :param type: Type of ABI element to filter by.
+    :type type: `str`
+    :param contract_abi: Contract ABI.
+    :type contract_abi: `ABI`
+    :return: List of ABI elements of the specified type.
+    :rtype: `Sequence[ABIElement]`
+
+    .. doctest::
+
+        >>> from eth_utils import filter_abi_by_type
+        >>> abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> filter_abi_by_type("function", abi)
+        [{'type': 'function', 'name': 'myFunction', 'inputs': [], 'outputs': []}, \
+{'type': 'function', 'name': 'myFunction2', 'inputs': [], 'outputs': []}]
+    """
+    return [abi for abi in contract_abi if abi["type"] == type]
+
+
+def get_all_function_abis(contract_abi: ABI) -> Sequence[ABIFunction]:
+    """
+    Return interfaces for each function in the contract ABI.
+
+    :param abi: Contract ABI.
+    :type abi: `ABI`
+    :return: List of ABIs for each function interface.
+    :rtype: `list[ABIFunction]`
+
+    .. doctest::
+
+        >>> from eth_utils import get_all_function_abis
+        >>> contract_abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> get_all_function_abis(contract_abi)
+        [{'type': 'function', 'name': 'myFunction', 'inputs': [], 'outputs': []}, \
+{'type': 'function', 'name': 'myFunction2', 'inputs': [], 'outputs': []}]
+    """
+    return [
+        cast(ABIFunction, function_abi)
+        for function_abi in filter_abi_by_type("function", contract_abi)
+    ]
+
+
+def get_all_event_abis(contract_abi: ABI) -> Sequence[ABIEvent]:
+    """
+    Return interfaces for each event in the contract ABI.
+
+    :param abi: Contract ABI.
+    :type abi: `ABI`
+    :return: List of ABIs for each event interface.
+    :rtype: `list[ABIEvent]`
+
+    .. doctest::
+
+        >>> from eth_utils import get_all_event_abis
+        >>> contract_abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> get_all_event_abis(contract_abi)
+        [{'type': 'event', 'name': 'MyEvent', 'inputs': []}]
+    """
+    return [
+        cast(ABIEvent, event) for event in filter_abi_by_type("event", contract_abi)
+    ]
+
+
+def get_normalized_abi_inputs(
+    abi_element: ABIElement,
+    args: Sequence[Any],
```

Comment:
It looks like `args` and `kwargs` can be optional, is that true? If so, we should update these.